[DependencyInjection] Added information about deprecated aliases in debug:autowiring

XuruDragon opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | none
| License       | MIT
| Doc PR        | n/a

Fix and improves a bit PR #29968 and #29995 

![Screenshot](https://i.imgur.com/GDj2NIY.png)

Commit:
[DependencyInjection] Added information about deprecated aliases in debug:autowiring

| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | none
| License       | MIT
| Doc PR        | n/a

Fix and improves a bit PR #29968 and #29995

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Alias.php:
```diff
@@ -125,6 +125,11 @@ public function getDeprecationMessage(string $id): string
         return str_replace('%alias_id%', $id, $this->deprecationTemplate ?: self::$defaultDeprecationTemplate);
     }
 
+    public function getDeprecationTemplate(): ?string
```
The same concept exists on `Definition` but this method doesn't exists there, which means it should be removed to me.

XuruDragon replied:
indeed that could be removed

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/CHANGELOG.md:
```diff
@@ -6,6 +6,7 @@ CHANGELOG
 
  * added `%env(trim:...)%` processor to trim a string value
  * added `%env(default:...)%` processor to fallback to a default value
+ * added support for deprecated aliases
```
s/deprecated/deprecating/

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Compiler/ResolveReferencesToAliasesPass.php:
```diff
@@ -34,7 +34,10 @@ public function process(ContainerBuilder $container)
+                $container->setAlias($id, $defId)
+                    ->setPublic($alias->isPublic())
+                    ->setPrivate($alias->isPrivate())
+                    ->setDeprecated($alias->isDeprecated(), $alias->getDeprecationTemplate());
```
why should a deprecated alias propagate to its referring aliases?

XuruDragon replied:
I'll reset that to the previous code

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php:
```diff
@@ -354,7 +354,7 @@ private function parseDefinition($id, $service, $file, array $defaults)
 
                 if ('deprecated' === $key) {
-                    $alias->setDeprecated(true, $value);
+                    $alias->setDeprecated(true, \is_bool($value) ? null : $value);
```
`true === $value ? null ? $value`?

nicolas-grekas replied:
Actually, the same code for services doesn't have this logic, so this change might not be needed.

XuruDragon replied:
Yes indeed, but if we have specified `deprecated: true` in the alias configuration, without this condition an error is thrown because we do not give a string containing `%alias_id%` but `true`...

nicolas-grekas reviewed (commented):
Thanks!
Can you please share a screenshot?
It'd be nice to add some tests also.

XuruDragon commented:
> Thanks!
> Can you please share a screenshot?
> It'd be nice to add some tests also.

Screenshot added.
Tests already existing, added by #29968, I've fixed one of them.

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php:
```diff
@@ -354,7 +354,7 @@ private function parseDefinition($id, $service, $file, array $defaults)
 
                 if ('deprecated' === $key) {
-                    $alias->setDeprecated(true, $value);
+                    $alias->setDeprecated(true, true === $value ? null : $value);
```
should be reverted I think - at least we don't do that for services, so why do it for aliases?

XuruDragon replied:
As I said in the last conversation :

Yes indeed, but if we specify `deprecated: true` in the alias configuration, without this condition an error is thrown because we do not give a string containing `%alias_id%` but `true`...

nicolas-grekas replied:
can you check if we support that for Definition also? because if yes, there is a bug to fix with them - and if not, there is a bug to fix with aliases :) A test case would help secure this.

XuruDragon replied:
I think there is a bug fix to do.
If we give `deprecated: true` in the configuration it crash like the alias configuration does.

I can do the same test than for alias to solve that

nicolas-grekas replied:
If we support `true`, we need to support `false` I suppose, not really meaningful to me.
It's not [in the doc](https://symfony.com/doc/current/service_container/alias_private.html#deprecating-services) but [in the blog post](https://symfony.com/blog/new-in-symfony-4-3-deprecating-service-aliases) - which is wrong and should be fixed. The [article in 2015](https://symfony.com/blog/new-in-symfony-2-8-deprecated-service-definitions) was correct.
Let's just remove that change and done, it already fails you told me.

nicolas-grekas replied:
blog post updated.

XuruDragon replied:
so we doesn't support true and I remove all that thing ? (test included)

nicolas-grekas replied:
Yep, `true` would be a new feature, thus unrelated - and unneeded to me.
If we can keep a test case ensuring there is a proper failure when `true` is passed, that'd be nice.

XuruDragon replied:
ok, i'll do the change

XuruDragon replied:
done

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/Command/DebugAutowiringCommand.php:
```diff
@@ -104,7 +104,12 @@ protected function execute(InputInterface $input, OutputInterface $output)
+                $serviceAlias = $builder->getAlias($serviceId);
+                $serviceLine .= ' <fg=cyan>('.$serviceAlias.')</>';
+
+                if ($serviceAlias->isDeprecated()) {
```
since `Alias::isDeprecated()` was added in master, we need to bump the minimum version of DI in FWB's composer.json to `^4.3`

XuruDragon replied:
i'll bump it to `^4.3`

nicolas-grekas commented:
Thank you @XuruDragon.
